src/Simulation/qdk_sim_rs/src/linalg/array_ext.rs

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//! Private extension traits and methods for working with arrays.
+
+use ndarray::{
+    Array, ArrayBase, Axis, Data, DataMut, Dimension, Ix2, OwnedRepr, RawData, RemoveAxis,
+};
+use num_traits::Zero;
+
+use crate::error::QdkSimError;
+
+pub(crate) trait ShapeExt {
+    fn require_square(&self) -> Result<usize, QdkSimError>;
+}
+
+impl<S, D> ShapeExt for ArrayBase<S, D>
+where
+    S: RawData,
+    D: Dimension,
+{
+    fn require_square(&self) -> Result<usize, QdkSimError> {
```

Comment:
This is a very general error type but there is only one possible error in this case. Maybe `Option<usize>`?

Reply:
My thinking here was that this is an internal-only trait/impl pair (`pub(crate)` in particular), and is mainly used to generate the specific error case `NotSquare` as a utility. Were this a public API, I'd agree entirely, but making that change here would necessitate adding a bunch of `.ok_or` calls where the return value for this extension method is used.